pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,665 @@
+	defaultMaxReplicaUpdateCount = 1
+	updateAttemptInterval        = time.Duration(1) * time.Millisecond
+	updateTimeout                = time.Duration(1) * time.Minute
+)
+
+type AttachmentProvisioner interface {
+	PublishVolume(ctx context.Context, volumeID string, nodeID string, volumeContext map[string]string) (map[string]string, error)
+	UnpublishVolume(ctx context.Context, volumeID string, nodeID string) error
+}
+
+type ReconcileAttachDetach struct {
+	client                client.Client
+	azVolumeClient        azVolumeClientSet.Interface
+	kubeClient            kubeClientSet.Interface
+	namespace             string
+	attachmentProvisioner AttachmentProvisioner
+}
+
+var _ reconcile.Reconciler = &ReconcileAttachDetach{}
+
+func (r *ReconcileAttachDetach) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, request.Name, request.Namespace, true)
+	// if object is not found, it means the object has been deleted. Log the deletion and do not requeue
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// detachment request
+	if requestedToBeDeleted(azVolumeAttachment) {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.Attached || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed || azVolumeAttachment.Status.State == v1alpha1.DetachmentFailed {
+			if err := r.triggerDetach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// attachment request
+	} else if azVolumeAttachment.Status.Detail == nil {
+		if azVolumeAttachment.Status.State == v1alpha1.AttachmentPending || azVolumeAttachment.Status.State == v1alpha1.AttachmentFailed {
+			if err := r.triggerAttach(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+		// promotion request
+	} else if azVolumeAttachment.Spec.RequestedRole != azVolumeAttachment.Status.Detail.Role {
+		if err := r.promote(ctx, azVolumeAttachment.Name, azVolumeAttachment); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileAttachDetach) triggerAttach(ctx context.Context, attachmentName string, azVolumeAttachment *v1alpha1.AzVolumeAttachment) error {
+	// initialize metadata and update status block
+	conditionFunc := func() (bool, error) {
+		azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, attachmentName, r.namespace, true)
+		if err != nil {
+			klog.Errorf("failed to get AzVolumeAttachment (%s): %v", attachmentName, err)
+			// if failed to get azVolumeAttachment then escape the wait
+			return false, err
+		}
+
+		// deepCopy
```

Comment:
This isn't a very useful comment since the next line is pretty obvious. Consider combining this and the following two comments into one, e.g.
```go
// Copy the object, add finalizer and label, and update the state to "Attaching".
```